src/crafting.cpp

```diff
@@ -42,268 +43,134 @@ recipe::recipe() :
-        if( ja.size() >= 3 ) {
-            bd.recipe_name = ja.get_string( 2 );
-            bd.hidden = bd.recipe_name.empty();
+    } else if( jo.has_array( "autolearn" ) ) {
+        r.autolearn = false;
+        auto sk = jo.get_array( "autolearn" );
+        while( sk.has_more() ) {
+            auto arr = sk.next_array();
+            r.autolearn_requirements[skill_id( arr.get_string( 0 ) )] = arr.get_int( 1 );
         }
-        rec->booksets.push_back( bd );
     }
 
-    // Note, a recipe has to be fully instantiated before adding
-    recipe_dict.add( rec );
-}
+    if( jo.has_member( "decomp_learn" ) ) {
+        r.learn_by_disassembly.clear();
 
-void reset_recipes()
-{
-    recipe_dict.clear();
-}
-
-void finalize_recipes()
-{
-    // Has to be done before recipe finalization!
-    for( auto r : recipe_dict ) {
-        r->requirements_ = std::accumulate( r->reqs.begin(), r->reqs.end(), requirement_data(),
-        []( const requirement_data & lhs, const std::pair<requirement_id, int> &rhs ) {
-            return lhs + ( *rhs.first * rhs.second );
-        } );
-    }
-
-    recipe_dict.finalize();
-
-    std::ostringstream buffer;
-    for( auto r : recipe_dict ) {
-        buffer.clear();
-        for( auto j = r->booksets.begin(); j != r->booksets.end(); ++j ) {
-            const std::string &book_id = j->book_id;
-            if( !item::type_is_defined( book_id ) ) {
-                buffer << "book " << book_id << " for recipe " << r->ident() << " does not exist" << "\n";
-                continue;
-            }
-            const itype *t = item::find_type( book_id );
-            if( !t->book ) {
-                // TODO: we could make up a book slot?
-                buffer << "book " << book_id << " for recipe " << r->ident() << " is not a book" << "\n";
-                continue;
+        if( jo.has_int( "decomp_learn" ) ) {
+            if( !r.skill_used ) {
+                jo.throw_error( "decomp_learn specified with no skill_used" );
             }
-            islot_book::recipe_with_description_t rwd{ r, j->skill_level, "", j->hidden };
-            if( j->recipe_name.empty() ) {
-                rwd.name = item::nname( r->result );
-            } else {
-                rwd.name = _( j->recipe_name.c_str() );
+            assign( jo, "decomp_learn", r.learn_by_disassembly[r.skill_used] );
+
+        } else if( jo.has_array( "decomp_learn" ) ) {
+            auto sk = jo.get_array( "decomp_learn" );
+            while( sk.has_more() ) {
+                auto arr = sk.next_array();
+                r.learn_by_disassembly[skill_id( arr.get_string( 0 ) )] = arr.get_int( 1 );
             }
-            t->book->recipes.insert( rwd );
         }
-        r->booksets.clear();
+    }
 
-        if( !item::type_is_defined( r->result ) ) {
-            buffer << "Recipe " << r->ident() << " defines invalid result " << r->result << "\n";
+    if( jo.has_member( "byproducts" ) ) {
```

Comment:
Prior to this PR there are four possible forms `byproducts` (although not all were used). It's very difficult to write automated tooling that doesn't struggle with that (there is a tendency to flip between two semantically equivalent forms).

If we are going to aim for better tooling (linters, editors etc) and still want to maintain ability to edit JSON directly (definitely important) then having a canonical form is important.